### PR TITLE
fontconfig: add option to use system directories

### LINKED ISSUE #5782: [package] fontconfig/2.13.93: How to use with sysconfdir set on build machine
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **fontconfig/2.13.93**
  * Operating System+version: **Linux Ubuntu 20.04**
  * Compiler+version: **GCC 9.3.0**
  * Conan version: **conan 1.36.0**
  * Python version: **Python 3.8.5**

### Conan profile
```
[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=9
compiler.libcxx=libstdc++11
build_type=Release
[options]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)
Currently I'm trying to build an app with Qt6. When I run the app I get the error 
```
Fontconfig error: Cannot load default config file: No such file: (null)
```
When investigating this with strace I found that fontconfig looks at 
```
/home/conan/w/BuildSingleReference/.conan/data/fontconfig/2.13.93/_/_/package/211285f7c5eb7915075b241bd6a053e345cec58c/bin/etc/fonts/fonts.conf
```

which seems to come from the way sysconfdir is defined in the recipe: https://github.com/conan-io/conan-center-index/blob/38d85e32c8352c7faabf191901ac126ec1214f95/recipes/fontconfig/all/conanfile.py#L61

Of course this behavior can be avoided by setting the environment variable `FONTCONFIG_PATH`.

As I want my app easily distributable through conan, what is the recommended way to hint fontconfig to the right path (e.g. `/etc/fonts` on Ubuntu)?

Thanks in advance!

### PR DESCRIPTION
**Pull request by johningve**
The fontconfig package currently tries to load fonts from some directory within the conan package directory. This results in the following error when this directory is not present:

```
Fontconfig error: Cannot load default config file: No such file: (null)
```

This commit adds an option `system_dirs` to use system directories instead.

Fixes #5782

Specify library name and version:  **fontconfig/2.13.93+**

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/fontconfig/meson/conanfile.py [hunk 1]
```diff
@@ -79,12 +82,23 @@ def generate(self):
...
+        build_script = os.path.join(self.source_folder, "meson.build")
+        replace_in_file(self, build_script,
                         "freetype_req = '>= 21.0.15'",
                         f"freetype_req = '{Version(self.dependencies['freetype'].ref.version)}'")
 
+        if self.settings.os in ("Linux", "FreeBSD") and self.options.get_safe("system_dirs", False):
+            self._replace_configuration_string(build_script, "CONFIGDIR", "fc_configdir", "'/etc/fonts/conf.d'")
+            self._replace_configuration_string(build_script, "FC_CACHEDIR", "fc_cachedir", "'/var/cache/fontconfig'")
+            self._replace_configuration_string(build_script, "FC_TEMPLATEDIR", "fc_templatedir", "'/usr/share/fontconfig/conf.avail'")
+            self._replace_configuration_string(build_script, "FONTCONFIG_PATH", "fc_baseconfigdir", "'/etc/fonts'")
```
Comment (valgur) [inline recipes/fontconfig/meson/conanfile.py:100]: You should be able to accomplish the same without patching simply via project_options in generate(), I think:
```
tc.project_options["fc_configdir"] = "/etc/fonts/conf.d"
tc.project_options["fc_cachedir"] = "/var/cache/fontconfig"
tc.project_options["fc_templatedir"] = "/usr/share/fontconfig/conf.avail"
tc.project_options["fc_baseconfigdir"] = "/etc/fonts"
```
Also, you should probably override `xml-dir` as well? https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/main/meson_options.txt?ref_type=heads#L32-45

**Review by valgur — CHANGES_REQUESTED** [review]
This fixes what seems like an important oversight. Thanks!

Should use `project_options` instead, though.

**Comment by valgur** [comment]
Could setting either of these options get the best of both worlds without introducing a recipe option, maybe?
https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/main/meson_options.txt?ref_type=heads#L25-29

**Comment by johningve** [comment]
Incidentally I noticed that the paths get different defaults on Conan 2 compared to Conan 1 (that is, when `system_dirs=False`). The Paths on Conan 2 are almost correct for a Linux system. Though the Template and XML directory paths are wrong.

Conan 2:

```
  Defaults
    Hinting                    : slight
    Font directories           : /usr/share/fonts, /usr/local/share/fonts
    Additional font directories:

  Paths
    Cache directory            : /var/cache/fontconfig
    Template directory         : /share/fontconfig/conf.avail
    Base config directory      : /etc/fonts
    Config directory           : /etc/fonts/conf.d
    XML directory              : /share/xml/fontconfig

```

Conan 1:

```
  Defaults
    Hinting                    : slight
    Font directories           : /usr/share/fonts, /usr/local/share/fonts
    Additional font directories:

  Paths
    Cache directory            : /home/johnio/.conan/data/fontconfig/2.14.2/_/_/package/91934fc8bcc809a17bd34af66d8491f41cb49e99/var/cache/fontconfig
    Template directory         : /home/johnio/.conan/data/fontconfig/2.14.2/_/_/package/91934fc8bcc809a17bd34af66d8491f41cb49e99/share/fontconfig/conf.avail
    Base config directory      : /home/johnio/.conan/data/fontconfig/2.14.2/_/_/package/91934fc8bcc809a17bd34af66d8491f41cb49e99/etc/fonts
    Config directory           : /home/johnio/.conan/data/fontconfig/2.14.2/_/_/package/91934fc8bcc809a17bd34af66d8491f41cb49e99/etc/fonts/conf.d
    XML directory              : /home/johnio/.conan/data/fontconfig/2.14.2/_/_/package/91934fc8bcc809a17bd34af66d8491f41cb49e99/share/xml/fontconfig

```

File: recipes/fontconfig/meson/conanfile.py (hunk 1 above)
Comment (johningve) [reply recipes/fontconfig/meson/conanfile.py:100]: I thought I had tried that previously, but you are right, it works on Conan 2.

Unfortunately, on Conan 1, Meson attempts to install something to /usr and fails. Looks like it is due to missing this change: https://github.com/conan-io/conan/pull/15706

**Comment by johningve** [comment]
> Could setting either of these options get the best of both worlds without introducing a recipe option, maybe? https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/main/meson_options.txt?ref_type=heads#L25-29

Looks like those options are used to specify the location of fonts, not the config/cache directories.

**Review by uilianries — CHANGES_REQUESTED** [review]
@johningve Thank you for your suggestion of fixing that case. I would not go for adding a new option only change an environment variable, it's too much. Conan offers mechanisms to do it without touching the recipe. Please, read https://github.com/conan-io/conan-center-index/issues/5782#issuecomment-2149686757

**Comment by johningve** [comment]
> @johningve Thank you for your suggestion of fixing that case. I would not go for adding a new option only change an environment variable, it's too much. Conan offers mechanisms to do it without touching the recipe. Please, read [#5782 (comment)](https://github.com/conan-io/conan-center-index/issues/5782#issuecomment-2149686757) and check if fits your case.

It is not about changing an environment variable, that is a workaround to this issue. What this is about is changing the  location where fontconfig will look for fonts by default. I want to be able to include fontconfig in my application and have it work without the user having to worry about environment variables. 

> I see a fragility here as you are hardcoding system paths that we can not guarantee. Conan only knows its own cache, outside of that is risky. Still, in case that option is not enough, we could add a Conan configuration to handle those variables, they fit better because are not part of the package ID.

I would say that at least on Linux, these paths are well known and relied upon. I get that Conan doesn't want to rely upon the outside too much, but the purpose of fontconfig is to enable applications to find and use the fonts available on the local system.

**Comment by uilianries** [comment]
Hello again!

I’m following up on the discussion regarding the default system directories used by fontconfig. I’ve provided a more detailed explanation of how to parameterize the system folders in my comment here: https://github.com/conan-io/conan-center-index/issues/5782#issuecomment-2161050812.

Thank you once again for your pull request and for proposing an alternative to address this limitation. However, the current implementation seems to introduce additional complexity to the Conan recipe. This includes modifying the package ID with a new recipe option, requiring several custom user configurations, and applying patches dynamically.

For these reasons, I’ll be closing this pull request, as I’m not inclined to accept the proposed changes. I encourage you to explore using Conan’s native configuration to extend the Meson configuration for this use case.

Please feel free to share your feedback or ask for help on how to use Conan configuration for your particular case.

Thank you for your understanding!

**Pull request closed without merging**